### PR TITLE
Convert post_shutdown_to_master to inlineCallbacks

### PR DESCRIPTION
**Pull request by opalmer**
This change converts post_shutdown_to_master to inlineCallbacks.  In addition this change also fixes a couple of bugs regarding the shutdown timeout.

### CODE REVIEW

**Comment by opalmer** [comment]
@guidow, PTAL.  I've finished adding tests for this.

**Comment by guidow** [comment]
With these patches, the agent fails to shut down when requested. It will successfully post the changed state to the master, including retries if the master was not available at first, but will then fail to actually shut down.

**Comment by opalmer** [comment]
Ok, that should be fixed now.  I had a feeling stop() may break but didn't get a chance to manually test it last night.  With this change stop() now just stops the agent instead of stopping the reactor too which is a little cleaner and it makes stop() more testable.  

I was planning to address convert `stop()` to inlineCallbacks in another branch but this seems like a good place to make the change since it's related.  But, I'm going to add tests for the new code before this merges (especially since stop()/sigint_handler() didn't have tests before)

**Comment by opalmer** [comment]
@guidow, PTAL. There's tests for `sigint_handler()` and `stop()` now.  I made a few code changes to `stop()` but it should help in catching a few possible conditions we didn't before.

File: pyfarm/agent/service.py [hunk 1]
```diff
@@ -472,151 +483,183 @@ def start(self, shutdown_events=True, http_server=True):
...
-                    reactor.stop()
+        svclog.info(
+            "Waiting on %s job types to terminate", len(config["jobtypes"]))
+
+        while True:
+            # No jobtypes remain
+            if not config["jobtypes"]:
+                break
+
+            # We're hit the timeout
```
Comment (guidow) [inline pyfarm/agent/service.py]: *We have

File: pyfarm/agent/service.py [hunk 2]
```diff
@@ -472,151 +483,183 @@ def start(self, shutdown_events=True, http_server=True):
...
-                    self.post_shutdown_to_master()
-                else:
-                    reactor.stop()
+        svclog.info(
+            "Waiting on %s job types to terminate", len(config["jobtypes"]))
+
+        while True:
+            # No jobtypes remain
+            if not config["jobtypes"]:
+                break
```
Comment (guidow) [inline pyfarm/agent/service.py]: Can't you rewrite the while loop instead as
```while config["jobtypes"] and datetime.utcnow() <= self.shutdown_timeout```?

A ```while True``` in a place that's not actually supposed to be an endless loop seems kinda wrong to me...

**Comment by guidow** [comment]
It works now.

File: pyfarm/agent/service.py (hunk 2 above)
Comment (opalmer) [inline pyfarm/agent/service.py]: Agreed, fixed.

**Comment by opalmer** [comment]
Thanks for the review, since the suggested change is not functional and it didn't break the tests I'm going to go ahead and merge this.

**Pull request merged**